From 4623b19a078e68b0679582a0e99793aba6dd0507 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 7 Aug 2015 10:14:51 +0800
Subject: [PATCH] bfs: v4.1_0463_1 rcu stall fix.

---
 kernel/sched/bfs.c | 10 ----------
 1 file changed, 10 deletions(-)

diff --git a/kernel/sched/bfs.c b/kernel/sched/bfs.c
index 609eab3..e3966fe 100644
--- a/kernel/sched/bfs.c
+++ b/kernel/sched/bfs.c
@@ -3844,10 +3844,6 @@ void rt_mutex_setprio(struct task_struct *p, int prio)
 
 	trace_sched_pi_setprio(p, prio);
 	oldprio = p->prio;
-
-	if (task_queued(p))
-		dequeue_task(p);
-
 	p->prio = prio;
 	update_task_priodl(p);
 
@@ -4256,9 +4252,6 @@ recheck:
 		return 0;
 	}
 
-	if (task_queued(p))
-		dequeue_task(p);
-
 	__setscheduler(rq, p, attr, true);
 
 	check_task_changed(rq, p, oldprio);
@@ -7280,9 +7273,6 @@ static void normalize_task(struct rq *rq, struct task_struct *p)
 	};
 	int old_prio = p->prio;
 
-	if (task_queued(p))
-		dequeue_task(p);
-
 	__setscheduler(rq, p, &attr, false);
 
 	check_task_changed(rq, p, old_prio);
-- 
2.1.1

